Split unpinned-uses into two separate checks

funnelfiasco opened the pull request:
Here's my first pass at implementing #179. The code is there and functional, AFAICT. I've tested it against a real workflow file that uses repo paths. I don't have any handy that use Docker, but if I understand the model correctly, it's very straightforward to check it and so my code will Just Work™.

~~I will add tests and docs before I remove the "WIP" from this, but I wanted to make sure what I've done here is generally acceptable before I put effort into that part.~~

~~Since most of the "what to do if we find an unpinned `uses` of either flavor" is repetitive, I tossed in a flag instead of repeating that code.~~

There's a lot of "not" in the new function I created in the model. It could have been written a little more cleanly without the negations, but that kept the usage the same as in the existing `unpinned()`, which seemed more important.

Of note, this only looks to see if the pin _looks like_ a SHA. It does not attempt to verify that such a SHA actually exists. You probably expect that, but I wanted to be explicit about it (and I will be in the docs updates, when I create them)

~~Edit to add: I have also not run fmt or clippy yet, since it seemed like something that could wait until I got a thumbs up on the general direction.~~

Fixes #179

funnelfiasco commented on src/models.rs:
```diff
@@ -433,6 +434,14 @@ impl<'a> Uses<'a> {
+    pub(crate) fn unpinned_sha(&self) -> bool {
+        let regex = Regex::new("^[a-fA-F0-9]{40}$").unwrap();
+        match self {
+            Uses::Docker(docker) => ! docker.hash.is_none(),
```
Clippy tells me that I could use `docker.has.is_some()` here instead, so I'll do that.

woodruffw commented on src/models.rs:
```diff
@@ -433,6 +434,14 @@ impl<'a> Uses<'a> {
+        let regex = Regex::new("^[a-fA-F0-9]{40}$").unwrap();
+        match self {
+            Uses::Docker(docker) => ! docker.hash.is_none(),
+            Uses::Repository(repo) => repo.git_ref.is_none() || ! regex.is_match(repo.git_ref.unwrap()),
```
If I'm following the logic here correctly, I think this can reuse the existing `ref_is_commit()` helper:

```suggestion
            Uses::Repository(repo) => !repo.ref_is_commit(),
```

(and then you can remove the `Regex` object :slightly_smiling_face:)

woodruffw commented on src/models.rs:
```diff
@@ -433,6 +434,14 @@ impl<'a> Uses<'a> {
         }
     }
+
+    pub(crate) fn unpinned_sha(&self) -> bool {
```
Bikeshed: `unpinned_sha` is a slightly confusing name for me, maybe this could be `unhashed` if the intended logic is "this `uses:` is either entirely unpinned or, if pinned, does not use a hashed pin"?

funnelfiasco replied:
I think that's a good color of paint, yes. I didn't like the initial name, but I wanted something that suggested similarity between the two. Your suggestion is a big improvement. I'll go with it unless something brilliant comes to mind before I sit down to work on this later today.

woodruffw commented on src/audit/unpinned_uses.rs:
```diff
@@ -21,15 +23,31 @@ impl WorkflowAudit for UnpinnedUses {
         };
 
-        if uses.unpinned() {
+        // Check twice, complain once
+        let mut found_unpinned = false;
+        let mut annotation = "";
+        if self.state.pedantic {
+            if uses.unpinned_sha() {
+                found_unpinned = true;
+                annotation = "action is not pinned to a hash ref";
+            }
+        } else {
+            if uses.unpinned() {
+                found_unpinned = true;
+                annotation = "action is not pinned to a tag, branch, or hash ref";
+            }
+        }
+
+        // If one of the checks failed, add a finding
+        if found_unpinned {
```
FWIW, I think the idiomatic Rust structure for this would be something like:

```rust
let annotation = if { ... } else if { ... } else { continue; };
```

which would avoid the need for the `found_unpinned` variable and the `mut`s. I can send a commit to this branch showing what I mean If I'm not explaining that well :slightly_smiling_face:

woodruffw replied:
Actually, thinking about this a bit more -- perhaps it makes sense to emit 2 separate findings here, one for the non-pedantic case and one for the pedantic case? My thinking is that `--pedantic` shouldn't change the behavior shape of findings, it should only be strictly additive to the findings set.

Thoughts? I think that would also simplify the code slightly, since there's be no need to parametrize the annotation.

funnelfiasco replied:
Hm. So in the pedantic case, `uses: some/action` would emit two findings? I don't think that's particularly useful as a user, but I might be missing a use case for it.

Maybe this is just an abuse of `--pedantic` because we don't have a `--strict` or the like. In this specific case, we're not adding an audit, really, just intensifying one that already exists.

Having typed that out, my weak preference is to keep it as-is, with the more idiomatic style you suggested.

woodruffw replied:
> Maybe this is just an abuse of `--pedantic` because we don't have a `--strict` or the like. In this specific case, we're not adding an audit, really, just intensifying one that already exists.

That's a good point. I don't have a super strong opinion or intuition here, so we can keep it as one finding for now and see what happens as I tweak the pedantic/sensitivity model 🙂

funnelfiasco replied:
Oh, here's another thought: could we use different severities for the two so the user can use `--min-severity` to decide how strict they want to be? This is essentially the sensitivity model discussion, IMO

So I'd put the sha check at informative or low, and the entirely unpinned at medium or high. My inclination is to err on the stricter side, since this is an easy avenue for supply chain attacks, and go with unhashed is low and unpinned is high, but I don't have a strong take on that.

woodruffw commented:
Thanks @funnelfiasco! The overall approach here LGTM; I left a couple of nits. If you want to run `cargo fmt` and `cargo clippy` internally then I can do a full review pass.

funnelfiasco commented:
Updated based on feedback:

* Used `is_some()` for the Docker hash instead of `! .is_none()`
* Renamed new function `unhashed`
* Dropped use of pedantic in favor of marking unpinned as medium severity and unhashed as low severity
* Used `ref_is_commit` and dropped regex
* Ran fmt and clippy

Todo, pending approval of these changes:

* Docs
* Tests

funnelfiasco commented:
Rested, tested, and ready for review

woodruffw commented:
Thanks! I'll do a review pass tomorrow.

woodruffw commented on src/audit/unpinned_uses.rs:
```diff
@@ -21,15 +21,26 @@ impl WorkflowAudit for UnpinnedUses {
         };
 
-        if uses.unpinned() {
+        let (annotation, severity) = if uses.unpinned() {
+            (
+                "action is not pinned to a tag, branch, or hash ref",
+                Severity::Medium,
+            )
+        } else if uses.unhashed() {
+            ("action is not pinned to a hash ref", Severity::Low)
+        } else {
+            ("", Severity::Informational)
+        };
```
Doing a `return` in the `else` case here should avoid the need for an "empty" state:

```suggestion
        let (annotation, severity) = if uses.unpinned() {
            (
                "action is not pinned to a tag, branch, or hash ref",
                Severity::Medium,
            )
        } else if uses.unhashed() {
            ("action is not pinned to a hash ref", Severity::Low)
        } else {
            return Ok(vec![]);
        };
```

woodruffw commented on src/audit/unpinned_uses.rs:
```diff
@@ -21,15 +21,26 @@ impl WorkflowAudit for UnpinnedUses {
+            ("", Severity::Informational)
+        };
+
+        if !annotation.is_empty() {
```
With the above, you _should_ be able to remove this `is_empty()` check:

```suggestion
```

funnelfiasco commented:
Updated based on feedback

woodruffw reviewed (approved):
Thanks, LGTM! I thought about it a bit more and I'm mildly worried that having the `unhashed()` check by default (vs. pedantic) is going to cause too much alert fatigue for most CI users. But that's not a blocker here, since this is a good impetus for me to rethink the sensitivity defaults/flags.
